### DISCUSSION TITLE
Tasmota Soft Serial Bridge Receive

### DISCUSSION THREAD

**Discussion by smhmaqsood** [arendst/Tasmota#12848] [Q&A] [answered]
Hi,

I'm willing to send and receive command through soft serial in tasmota serial bridge, I have defined TX and RX pins. Data is being transmitted successfully but I'm unable to receive data. I have tried to verify it through hardware serial/ console/ MQTT but unable to read received data or maybe data is not receiving. Need help in receiving the data on soft serial and make it able to send through MQTT.

Thanks

**Comment by Jason2866** [comment]
What happens if you connect TX to RX gpios (you have defined) from the device and send a message?

**Reply by smhmaqsood** [reply]
It is sending a message and I can receive that on second uart. but when i send string from second uart to tasmota then it doesn't show any data like {"SSerialReceived":"<string>"}

**Reply by smhmaqsood** [reply]
> 
> 
> What happens if you connect TX to RX gpios (you have defined) from the device and send a message?

I tried by shorting the soft TX & RX of tasmota but still receiving no data on serial bridge of tasmota.

**Comment by Jason2866** [comment]
Please post your exact config.

**Reply by smhmaqsood** [reply]
> 
> 
> Please post your exact config.

I'm using ESP8266 and I'm using GPIO 13 as SerBr TX and GPIO 14 as SerBr RX in Generic Module.
When I try to send data (string) from MQTT then SerBr TX works properly with command "SSerialSend" and it can be seen on other UART.
But when I try to send data (string) from another UART to SerBr RX then no data is received on SerBr RX. 
To verify pins, I swapped the pins but still facing the same issue.
The main issue is in receiving the data.

**Reply by ortegafernando** [reply]
join GPIO13 and GPIO14 and "forget the other UART", use SSerialSend. Check if you recieve data in RX pin.

**Reply by smhmaqsood** [reply]
I've tried that too but still no data is received.

**Comment by Jason2866** [comment, marked as answer]
Can not reproduce. Works as expected. Tried example https://tasmota.github.io/docs/Rules/#control-relays-via-serial
with defined dummy relais

and get:
```
13:32:04.438 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"on"}
13:32:04.452 RUL: SSERIALRECEIVED#DATA=ON performs "Power1 1"
13:32:04.459 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"ON"}
13:32:04.464 MQT: stat/sonoff-D4407C/POWER1 = ON
13:32:06.655 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"off"}
13:32:06.672 RUL: SSERIALRECEIVED#DATA=OFF performs "Power1 0"
13:32:06.679 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"OFF"}
13:32:06.684 MQT: stat/sonoff-D4407C/POWER1 = OFF
```

**Reply by smhmaqsood** [reply]
I don't want to use any sort of rules. All I want is to simply read the data on soft serial. 
I'm able to receive and send the data on hardware serial but on software serial I am only able to send the data and unable to receive it.

**Reply by Jason2866** [reply]
The rule was only to demonstrate that data is correctly received via serial as can be seen `tele/sonoff-D4407C/RESULT = {"SSerialReceived":"on"}` `tele/sonoff-D4407C/RESULT = {"SSerialReceived":"off"}`

**Reply by smhmaqsood** [reply]
Ok then why is it not working without rules

**Reply by smhmaqsood** [reply]
> 
> 
> Can not reproduce. Works as expected. Tried example https://tasmota.github.io/docs/Rules/#control-relays-via-serial
> with defined dummy relais
> 
> and get:
> 
> ```
> 13:32:04.438 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"on"}
> 13:32:04.452 RUL: SSERIALRECEIVED#DATA=ON performs "Power1 1"
> 13:32:04.459 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"ON"}
> 13:32:04.464 MQT: stat/sonoff-D4407C/POWER1 = ON
> 13:32:06.655 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"off"}
> 13:32:06.672 RUL: SSERIALRECEIVED#DATA=OFF performs "Power1 0"
> 13:32:06.679 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"OFF"}
> 13:32:06.684 MQT: stat/sonoff-D4407C/POWER1 = OFF
> ```

I tried to replicate it but its still not responding

**Reply by smhmaqsood** [reply]
> 
> 
> Can not reproduce. Works as expected. Tried example https://tasmota.github.io/docs/Rules/#control-relays-via-serial
> with defined dummy relais
> 
> and get:
> 
> ```
> 13:32:04.438 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"on"}
> 13:32:04.452 RUL: SSERIALRECEIVED#DATA=ON performs "Power1 1"
> 13:32:04.459 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"ON"}
> 13:32:04.464 MQT: stat/sonoff-D4407C/POWER1 = ON
> 13:32:06.655 MQT: tele/sonoff-D4407C/RESULT = {"SSerialReceived":"off"}
> 13:32:06.672 RUL: SSERIALRECEIVED#DATA=OFF performs "Power1 0"
> 13:32:06.679 MQT: stat/sonoff-D4407C/RESULT = {"POWER1":"OFF"}
> 13:32:06.684 MQT: stat/sonoff-D4407C/POWER1 = OFF
> ```

```
16:07:23.350 MQT: Connected
16:07:23.353 MQT: tele/UTI_73D6A4/LWT = Online (retained)
16:07:23.355 MQT: cmnd/UTI_73D6A4/POWER = 
16:07:23.361 MQT: tele/UTI_73D6A4/INFO1 = {"Info1":
16:07:23.375 MQT: tele/UTI_73D6A4/INFO2 = {"Info2": 
16:07:23.385 MQT: tele/UTI_73D6A4/INFO3 = {"Info3":
16:07:23.393 MQT: stat/UTI_73D6A4/RESULT = {"POWER":"OFF"}
16:07:23.395 MQT: stat/UTI_73D6A4/POWER = OFF
16:07:24.162 QPC: Reset
16:07:27.180 MQT: tele/UTI_73D6A4/STATE = {"Time":"2021-08-13T16:07:27","Uptime":"0T00:00:11","UptimeSec":11,"Heap":30,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Laptop","BSSId":"F6:06:69:9E:41:07","Channel":11,"RSSI":100,"Signal":-41,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:07:30.962 CMD: rule
16:07:30.969 MQT: stat/UTI_73D6A4/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":92,"Free":419,"Rules":"ON SSerialReceived#Data=on DO Power1 1 ENDON   ON SSerialReceived#Data=off DO Power1 0 ENDON"}}
16:07:37.162 MQT: tele/UTI_73D6A4/STATE = {"Time":"2021-08-13T16:07:37","Uptime":"0T00:00:21","UptimeSec":21,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Laptop","BSSId":"F6:06:69:9E:41:07","Channel":11,"RSSI":100,"Signal":-45,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:07:43.799 CMD: SSerialSend On
16:07:43.810 MQT: stat/UTI_73D6A4/RESULT = {"SSerialSend":"Done"}
16:07:47.137 MQT: tele/UTI_73D6A4/STATE = {"Time":"2021-08-13T16:07:47","Uptime":"0T00:00:31","UptimeSec":31,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Laptop","BSSId":"F6:06:69:9E:41:07","Channel":11,"RSSI":100,"Signal":-39,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:07:57.172 MQT: tele/UTI_73D6A4/STATE = {"Time":"2021-08-13T16:07:57","Uptime":"0T00:00:41","UptimeSec":41,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":25,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Laptop","BSSId":"F6:06:69:9E:41:07","Channel":11,"RSSI":100,"Signal":-41,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:08:07.154 MQT: tele/UTI_73D6A4/STATE = {"Time":"2021-08-13T16:08:07","Uptime":"0T00:00:51","UptimeSec":51,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Laptop","BSSId":"F6:06:69:9E:41:07","Channel":11,"RSSI":100,"Signal":-39,"LinkCount":1,"Downtime":"0T00:00:05"}}
```



I tried to replicae it but it is still not working
I shorted the Tx & Rx pins

**Reply by sfromis** [reply]
How did you configure the Tx/Rx pins in Tasmota? When I config them as `SerBr Tx/Rx` I get the expected `SSerialReceived` returned when I do the `SSerialSend On`.

**Comment by Jason2866** [comment]
But SSerialsend from device does not work

**Comment by salvadorAnt** [comment]
How would be the topic to send commands via mqtt, and get it in tasmota serial port?

**Reply by sfromis** [reply]
Via MQTT, you can [send any Tasmota command](https://tasmota.github.io/docs/MQTT/#commands-over-mqtt), and the topic to subscribe to is directly visible in the console with each JSON payload having `SerialReceived` data.

**Reply by salvadorAnt** [reply]
Sorry for poor english, I want to send a string to serial device through mqtt to Tasmota in which are configured serial bridge pins.
I can receive serial data from device in a payload, but i don't know how would be topic/payload to send it a command.

**Reply by sfromis** [reply]
Well, "any Tasmota command" includes the relevant serial send command, same command as you can use in the console, where this thread already has examples. And the docs page I linked clearly shows how the topics for that works.
